hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/storage/HoodieHFileWriter.java

```diff
@@ -122,7 +128,17 @@ public boolean canWrite() {
 
   @Override
   public void writeAvro(String recordKey, IndexedRecord object) throws IOException {
-    byte[] value = HoodieAvroUtils.avroToBytes((GenericRecord)object);
+    byte[] value = HoodieAvroUtils.avroToBytes((GenericRecord) object);
```

Comment:
I replied in the comment on why this has to be done by the other way.

The passed in object cannot be modified in this writeAvro() method. The caller if at all uses the record again will find the key missing and crash when dereferencing it. So, we need to empty the key and restore back the original key in the record so that the caller is not affected. Please take a look at the latest commit on the fix for this.